internal/service/handlers/activate_balance.go

```diff
@@ -0,0 +1,147 @@
+		return
+	}
+
+	balance, err := BalancesQ(r).FilterByNullifier(nullifier).Get()
+	if err != nil {
+		Log(r).WithError(err).Error("Failed to get balance by nullifier")
+		ape.RenderErr(w, problems.InternalError())
+		return
+	}
+
+	if balance == nil {
+		Log(r).Debugf("Balance %s not exist", nullifier)
+		ape.RenderErr(w, problems.NotFound())
+		return
+	}
+
+	if balance.ReferredBy != nil {
+		Log(r).Infof("Balance already activated with code '%s'", *balance.ReferredBy)
+		ape.RenderErr(w, problems.Conflict())
+		return
+	}
+
+	referral, err := ReferralsQ(r).FilterInactive().Get(req.Data.Attributes.ReferredBy)
+	if err != nil {
+		Log(r).WithError(err).Error("Failed to get referral by ID")
+		ape.RenderErr(w, problems.InternalError())
+		return
+	}
+	if referral == nil {
+		Log(r).Debugf("Referral code '%s' not found", referralCode)
+		ape.RenderErr(w, problems.NotFound())
+		return
+	}
+
+	refBalance, err := BalancesQ(r).FilterByNullifier(referral.Nullifier).Get()
+	if err != nil || refBalance == nil { // must exist due to FK constraint
+		Log(r).WithError(err).Error("Failed to get referrer balance by nullifier")
+		ape.RenderErr(w, problems.InternalError())
+		return
+	}
+
+	err = EventsQ(r).Transaction(func() error {
+		level, err := doLvlUpAndReferralsUpdate(Levels(r), ReferralsQ(r), *balance, 0)
+		if err != nil {
+			return fmt.Errorf("failed to do lvlup and referrals update: %w", err)
+		}
+
+		err = BalancesQ(r).FilterByNullifier(balance.Nullifier).Update(map[string]any{
+			data.ColReferredBy: referralCode,
+			data.ColLevel:      level,
+		})
+		if err != nil {
+			return fmt.Errorf("update balance amount and level: %w", err)
+		}
+
+		if refBalance.ReferredBy != nil {
+			err = EventsQ(r).Insert(data.Event{
+				Nullifier: nullifier,
+				Type:      models.TypeBeReferred,
+				Status:    data.EventFulfilled,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to insert `be_referred` event: %w", err)
+			}
+		}
+
+		if err = ReferralsQ(r).Consume(referralCode); err != nil {
+			return fmt.Errorf("failed to consume referral: %w", err)
+		}
+
+		evTypeRef := EventTypes(r).Get(models.TypeReferralSpecific, evtypes.FilterInactive)
+		if evTypeRef == nil {
+			Log(r).Debug("Referral specific event type is inactive")
+			return nil
+		}
+
+		if balance.IsVerified {
+			// Be referred event is a welcome bonus when you created balance with non-genesis referral code
+			if err = claimBeReferredEvent(r, *balance); err != nil {
+				return fmt.Errorf("failed to claim be referred event: %w", err)
+			}
+		}
+
+		// Adds a friend event for the referrer. If the event
+		// is inactive, then nothing happens. If active, the
+		// fulfilled event is added and, if possible, the event claimed
+		if err = addEventForReferrer(r, evTypeRef, *balance); err != nil {
+			return fmt.Errorf("add event for referrer: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		Log(r).WithError(err).Error("failed to insert events and consume referral for balance")
```

Comment:
General recommendation: start log messages from capital letter for consistency in service logs